src/server/watchdog.ts

```diff
@@ -0,0 +1,137 @@
+
+const log = (s: string) => {
+  process.stderr.write(`<watchdog> ${s}\n`);
+};
+
+// Disable signals - watchdog will self-destruct on input stream end.
+const doNothing = (signal: any) => {
+  log(`received ${signal}, ignoring`);
+};
+process.on('SIGINT', doNothing);
+process.on('SIGTERM', doNothing);
+
+process.on('uncaughtException', error => log(`unhandled exception: ${error.stack}`));
+process.on('unhandledRejection', reason => log(`unhandled rejection: ${reason}`));
+
+import * as childProcess from 'child_process';
+import * as fs from 'fs';
+import * as removeFolder from 'rimraf';
+
+const [,, executable, closeMethod, tempDirCount, ...rest] = process.argv;
+const tempDirs = rest.slice(0, +tempDirCount);
+const args = rest.slice(+tempDirCount);
+
+const cleanup = () => {
+  for (const dir of tempDirs) {
+    try {
+      removeFolder.sync(dir);
+    } catch (e) {
+      log(`error removing ${dir}: ${e.stack}`);
+    }
+  }
+};
+
+const stdioParam: ('inherit' | 'pipe')[] = closeMethod === 'none' ? ['pipe', 'inherit', 'inherit'] : ['inherit', 'inherit', 'inherit', 'pipe', 'inherit'];
+const spawnedProcess = childProcess.spawn(
+    executable,
+    args,
+    {
+      // On non-windows platforms, `detached: true` makes child process a leader of a new
+      // process group, making it possible to kill child process tree with `.kill(-pid)` command.
+      // @see https://nodejs.org/api/child_process.html#child_process_options_detached
+      detached: process.platform !== 'win32',
+      stdio: stdioParam,
+    }
+);
+// Prevent unhandled 'error' event.
+spawnedProcess.on('error', () => {});
+
+if (!spawnedProcess.pid) {
+  spawnedProcess.once('error', error => {
+    cleanup();
+    process.stderr.write(String(error.message));
+    process.exit(1);
+  });
+}
+process.stdout.write(`browser pid: ${spawnedProcess.pid}\n`);
+
+let timerId: NodeJS.Timeout | undefined;
+spawnedProcess.on('exit', (exitCode, signal) => {
+  // Do not unnecessary wait for 30 seconds after the browser did exit.
+  if (timerId)
+    clearTimeout(timerId);
+  log(`browser process did exit, cleaning up`);
+  cleanup();
+
+  // Exit with the same exit code.
+  if (exitCode !== null)
+    process.exit(exitCode);
+
+  // Here we mimic the signal of the browser process: removing our signal handlers
+  // and sending the signal to our own process, so that default handler does the magic.
+  process.removeListener('SIGINT' as any, doNothing);
+  process.removeListener('SIGTERM' as any, doNothing);
+  process.kill(process.pid, signal!);
+});
+
+const inputStream = closeMethod === 'none' ? process.stdin : fs.createReadStream('', {fd: 3});
+const outputStream = closeMethod === 'none' ? spawnedProcess.stdin : (spawnedProcess.stdio as any)[3] as NodeJS.WritableStream;
+// Note: we cannot use 'pipe' here since it prohibits our own 'write' to close the browser.
+inputStream.on('data', data => outputStream.write(data));
```

Comment:
I think [`stream.pipe`](https://nodejs.org/api/stream.html#stream_readable_pipe_destination_options) is warranted here. I don't fully understand what it does, but I think it's smarter than "on data write data".